### ISSUE TITLE
equation solve issue in Matlab

### ISSUE THREAD

**Issue by OsellaG** [cbm755/octsympy#913]
Hi,
I'm trying to use octsympy in matlab  (Octave is OK but I need some built-in matlab functions)  but I get some problem.

Some parts of the tool work fine but I cannot put at work the function solve.
syms x
solve(x^2,x)
give me the following error message:

Index exceeds matrix dimensions.

Error in python_ipc_sysoneline (line 146)
  A = extractblock(out(ind(1):end));

Error in python_ipc_driver (line 63)
        [A, db] = python_ipc_sysoneline(what, cmd, false, varargin{:});

Error in python_cmd (line 140)
  [A, db] = python_ipc_driver('run', cmd, varargin{:});

Error in sym/solve (line 152)
    out = python_cmd (cmd, varargin{:});

Anyone can help me?
Thanks in advance for any help.

**Comment by OsellaG** [comment]
The problem is in the low level Python command management

the tests__python_cmd.m stop with sintax error at ther following line

 r = python_cmd ('return 42 if sys.version_info >= (3,0) else long(42)');

**Comment by cbm755** [comment]
That's strange, that's an error from Python not Matlab.

What does `sympref diagnose` output?

My best guess here is that the PATH is different on Matlab and its running a different Python than is run on Octave.

**Comment by mtmiller** [comment]
If the syntax error is referring to the `return 42 if …` syntax, then yeah, it seems like your environment is using Python 2.4 or something older, which is not supported by octsympy.

You can check what Python is being used with

    system([sympref('python') ' --version']);

@cbm755 is that the best way to check this? I don't see `sys.version` reported in any verbose output or diagnostics anymore.

**Comment by cbm755** [comment]
> is that the best way to check this?

Can also do:
```
>> s = python_cmd('return sys.version')
s = 2.7.15 (default, Oct 15 2018, 15:24:06) 
[GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

I agree that `sympref diagnose` should be reporting the Python version.

**Comment by cbm755** [comment]
Are there two different things going on here?  I cannot reproduce either one.  It looks in Comment 2 like Python 2.4 is being run.  But then later comments (https://github.com/cbm755/octsympy/issues/913#issuecomment-434587134) show Python 3.7.1 and different errors...

@OsellaG, can you report the output of `s = python_cmd('return sys.version')`?

**Comment by cbm755** [comment]
Maybe you can paste the output of `sympref diagnose` as well?  Put it between three back tics like:
````
```
paste stuff here
```
````

**Comment by mtmiller** [comment]
If the Python version is in fact 3.7.1, then is it possible the program text being sent to Python is corrupted or truncated? Is there a way to save the intermediate program text in the sysoneline driver?

**Comment by cbm755** [comment]
Yeah, something like:
```diff
-  bigs = [headers s1 s2 s3];
+  bigs = [headers s1 s2 s3]
```
Can also try `sympref ipc systmpfile`...  IIRC, that creates temporary files with all the commands being used in the communication.

@OsellaG what OS are you on?  I don't think I've every tested Matlab + Windows.  And I've just noticed the default IPC on Windows is `sysoneline` not `system`...

Would still like to see full `sympref diagnose` output and to confirm which of the bugs you're able to reproduce.

**Comment by OsellaG** [comment]
Hi,
My OS is Win7 and MATLAB 2016b.
Sympref gives no errors (I'll attach the output on Monday).
I've already checked the "bigs" string passed to python: it's not corrupted when the python interpreter is called but the errors thrown by python is a syntax due to a truncated line.
I'll put more details on Monday.
P.s. the part of the script that write intermediate data to file doesn't work on MATLAB: it use fputs instead of fprintf.
Ciao

**Comment by OsellaG** [comment]
Hi everibody,
First of all the output of sympref diagnose (it's ok for me)
```
Symbolic package diagnostics
============================
Python and SymPy are needed for most features of the Symbolic package.
The Python interpreter is currently: "python".
Computers may have more than one Python interpreter installed.  If you
need to, you can select a different one using the PYTHON environment
variable (see "help sympref").  For example, to use Python 3, try
    setenv PYTHON python3
    sympref reset
Attempting to run python -c "print(\"Python says hello\")"


status =

     0


output =

Python says hello


Good, Python ran correctly.
SymPy Python Library
--------------------
SymPy is a Python library used by Symbolic for almost all features.
Attempting to run python -c "import sympy; print(sympy.__version__)"


status =

     0


output =

1.3


SymPy must be at least version 1.0; you have version 1.3.
We cannot automatically compare versions on Matlab: please verify above.
Good, a working version of SymPy is installed.
Python XML Parsing and DOM Support
----------------------------------
The XML DOM library is used by Symbolic for passing values to and from Python.
Some older versions of Python formatted XML output differently.  As long as you
have any reasonably recent version of Python, this should pass.
Attempting to run python -c "import xml.dom.minidom as minidom; doc = minidom.parseString(\"<item>value</item>\"); print(doc.toprettyxml(indent=\"\"))"


status =

     0


output =

<?xml version="1.0" ?>
<item>value</item>




Your kit looks good for running the Symbolic package.  Happy hacking!
```
e,g.
```
>> syms y
>> y^2-2*x+1
ans = (sym)

          2    
  -2*x + y  + 1
```

**Comment by OsellaG** [comment]
let's try something more  complex extract from test__python_cmd.m

```
r = python_cmd ('return 42 if sys.version_info >= (3,0) else long(42)')
Symbolic pkg v2.7.1: using one-line system() communication with SymPy.
Warning: this will be *SLOW*.  Every round-trip involves executing a
new Python process and many operations involve several round-trips.
Warning: "sysoneline" will fail when using very long expressions.
114   if (~mktmpfile)

status =

     1


out =

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 4
    return 42 if sys.version_info ,0) else long(42)
                                  ^
SyntaxError: invalid syntax



ind =

     []

Error using python_ipc_sysoneline (line 142)
sysoneline ipc: system() call failed!

Error in python_ipc_driver (line 63)
        [A, db] = python_ipc_sysoneline(what, cmd, false, varargin{:});

Error in python_cmd (line 140)
  [A, db] = python_ipc_driver('run', cmd, varargin{:});
```

the bigs variables is fprintf(bigs) 
```
fprintf(bigs)
exec(open("D:\OctavePKG\octsympy\private\python_header.py").read()); exec("try:
    _ins = []
    # end of a list
    octoutput_drv(\"PYTHON: successful variable import\")
except:
    echo_exception_stdout(\"while copying variables to Python\")
    raise"); exec("def _fcn(_ins):
    _outs = []
    try:
        return 42 if sys.version_info >= (3,0) else long(42)
    except Exception as e:
        ers = type(e).__name__ + \": \" + str(e) if str(e) else type(e).__name__
        _outs = (\"COMMAND_ERROR_PYTHON\", ers, sys.exc_info()[-1].tb_lineno)
    return _outs
_outs = _fcn(_ins)"); exec("try:
    octoutput_drv(_outs)
except:
    echo_exception_stdout(\"while copying variables from Python\")
    raise");
```

My opinion is that there is some problem in the python command:  bigs = [headers s1 s2 s3];
if I try system([pyexec '  -c "' headers s1 '"']) it work but if I try
system([pyexec '  -c "' headers s1 s2'"'])
I get an error from python.

**Comment by cbm755** [comment]
those `exec("try:` bits certainly look broken.  First one should be on a new line...

But ATOH this string is escaped so much I don't think I trust fprintf on it.  Can you use `disp` to show `bigs`?

**Comment by cbm755** [comment]
wait, maybe I mean none of it should be on newlines, the point is it needs to be all on one line (hence the name `sysoneline`).  At any rate, I don't trust `fprintf` on it.

**Comment by OsellaG** [comment]
I used fprintf to show how it will be in a command file like .sh (that doesn't work in WIN of course) but the command bigs is on one line.

```
disp(bigs)
exec(open(\"D:\\OctavePKG\\octsympy\\private\\python_header.py\").read()); exec(\"try:\n    _ins = []\n    # end of a list\n    octoutput_drv(\\\"PYTHON: successful variable import\\\")\nexcept:\n    echo_exception_stdout(\\\"while copying variables to Python\\\")\n    raise\"); exec(\"def _fcn(_ins):\n    _outs = []\n    try:\n        return 42 if sys.version_info >= (3,0) else long(42)\n    except Exception as e:\n        ers = type(e).__name__ + \\\": \\\" + str(e) if str(e) else type(e).__name__\n        _outs = (\\\"COMMAND_ERROR_PYTHON\\\", ers, sys.exc_info()[-1].tb_lineno)\n    return _outs\n_outs = _fcn(_ins)\"); exec(\"try:\n    octoutput_drv(_outs)\nexcept:\n    echo_exception_stdout(\\\"while copying variables from Python\\\")\n    raise\");
```

**Comment by cbm755** [comment]
Ah! Probably this is about those `>` characters doing stream redirection.  They may need to be escaped.

And that is why its failing at `>= (3,0)`.

But it seems we don't hit this on Octave on Windows with `sysoneline` (need to double-check), only on Matlab and Windows...?

**Comment by cbm755** [comment]
`r = python_cmd('return 10 > 2')` should be a minimal failing example?

**Comment by OsellaG** [comment]
Octave+Win7 are OK, don't show the problem.

**Comment by OsellaG** [comment]
something like this?

``
function s = myesc(s)

  for i = 1:length(s)
    % order is important here

    % escape quotes twice
    s{i} = strrep(s{i}, '\', '\\\\');

    % dbl-quote is rather special here
    % /" -> ///////" -> ///" -> /" -> "
    s{i} = strrep(s{i}, '"', '\\\"');
    
    %GC Escape sequence oly for Matlab
    s{i} = strrep(s{i}, '>', '\x3e');
    s{i} = strrep(s{i}, '<', '\x3c');

  end
end
` `

**Comment by OsellaG** [comment]
3)

`%test
 % cmd w/ backslash and \n must be escaped by user
 expy = 'a\b\\c\nd\';
 y = python_cmd ('return "a\\b\\\\c\\nd\\",');
octassert (strcmp(y, expy))
expy =

a\b\\c\nd\

y =

a\\b\\\\c\\nd\

`
4)

`%test
 % slashes
 x = '/\\ // \\\\ \\/\\/\\';
 z = '/\ // \\ \/\/\';
 y = python_cmd ('return _ins', x);
octassert (strcmp(y, x))
x =

/\\ // \\\\ \\/\\/\\

y =

/\\\\ // \\\\\\\\ \\\\/\\\\/\\

z =

/\ // \\ \/\/\
`
5)
`%test
 % slashes
 z = '/\ // \\ \/\/\';
 y = python_cmd ('return "/\\ // \\\\ \\/\\/\\"');
octassert (strcmp(y, z))
z =

/\ // \\ \/\/\

y =

/\\ // \\\\ \\/\\/\
`

**Comment by OsellaG** [comment]
6)

`%test
 % unicode
 s1 = 'æˆ‘çˆ±ä½ ';
 cmd = 'return u"\u6211\u7231\u4f60",';
 s2 = python_cmd (cmd);
octassert (strcmp (s1, s2))
s1 =

æˆ‘çˆ±ä½ 


s2 =

\u6211\u7231\u4f60

`

7)
`%test
 % unicode with \x
 s1 = 'æˆ‘';
 cmd = 'return b"\xe6\x88\x91".decode("utf-8")';
 s2 = python_cmd (cmd);
 octassert (strcmp (s1, s2))
s1 =

æˆ‘

s2 =

\xe6\x88\x91

`

8)
`fprintf('.')
%test
 % unicode with \x and some escaped backslashes
 s1 = '\æˆ‘\';
 cmd = 'return b"\\\xe6\x88\x91\\".decode("utf-8")';
 s2 = python_cmd (cmd);
octassert (strcmp (s1, s2))

s1 =

\æˆ‘\

s2 =

\\\xe6\x88\x91\
`
That's All

**Comment by cbm755** [comment]
Fix looks reasonable...  See also #871 and #874.  Can I interest you in doing a pull request?  Is it safe to do the fix on Octave/Win or do we need to protect it behind a OCTAVE_VER test?

Also, are you sure this works on Octave/Win?
```
sympref ipc sysoneline
r = python_cmd('return 10 > 2')
```
If its a problem with `system` and given that #871 happened on Octave/Win, I'm surprised this is Matlab specific...

Re: your other posts.  It seems there are some other escape differences in Matlab/Win versus Octave/Win!

**Comment by OsellaG** [comment]
The problem is the same in Octave-Win.
The fix works in Octave.
I can pull a request by myself? I'm quite new in GitHub usage.
I'm quite expert in matlab but not in python. :)

**Comment by cbm755** [comment]
Aha!  That's great news, makes it more reproducible.

I would be very happy if you do a pull request.  It seems complicated at first, but isn't too bad.  First "fork" the project here on Github, then "git clone" the development source code ("master") onto your computer.  Make a branch called "myfix" or whatever.  Do your chance and push the branch back to Github.  Lots of guides go into more detail.  Our https://github.com/cbm755/octsympy/wiki/Development-Guide isn't very good; I should add some links...


Take a look at 0c3797ead9255316cfd918e246f760d5f6bfd450: is that better/worse/the same as the fix you propose?

**Comment by OsellaG** [comment]
Yeah, is the same fix written in different way and in different place in the code.

**Comment by cbm755** [comment]
Thanks!  All things equal, maybe I prefer this `^>` to the `\x` stuff (at least its consistent throughout the code).

**Comment by OsellaG** [comment]
I Agree.

**Comment by cbm755** [comment]
Fixed by PR #929.

@OsellaG: next time, put the string `Fixes #913.` in the body of your git commit message and it github will close the issue automatically.
